### PR TITLE
refactor: basic element without type should not create union type

### PR DESCRIPTION
**Pull request by lizaiv77**
This PR has refactoring for basic types mainly addressing code reusability and using JSON tests instead mocks.
Generation wise what changed: if basic element does not have type like so
```json
{
   "name" : "element",
   "selector" : {}
}
```
Was: we still would create union type that would just extend BasicElement, now - we return BasicElement
```java
final BasicElement getElement() {
    return element(this.element).build(BasicElement.class, BasePageElement.class);
}
```
Consequently we do not declare interface type for element that is only used as a scope. 
Additional changes:
- moved Union types classes to separate package utam.compiler.types
- merged classes UnionTypeImpl and BasicElementUnionType
- discovered that we don't validate duplicate args in getter, added MethodParametersTracker and tests

### CODE REVIEW

File: utam-compiler/src/test/java/utam/compiler/grammar/UtamMethodTests.java [hunk 1]
```diff
@@ -40,107 +32,16 @@ private static void test(String jsonFile, String expectedError) {
...
-   */
-  @Test
-  public void testGetAbstractMethod() {
-    TranslationContext context = TestUtilities.getTestTranslationContext();
-    UtamMethod method = TestUtilities.UtamEntityCreator
-        .createUtamMethod(METHOD_NAME, "string", null);
-    PageObjectValidationTestHelper.MethodInfo info =
-        new PageObjectValidationTestHelper.MethodInfo(METHOD_NAME, "String");
-    PageObjectValidationTestHelper.validateMethod(method.getAbstractMethod(context), info);
-  }
```
Comment (lizaiv77) [inline utam-compiler/src/test/java/utam/compiler/grammar/UtamMethodTests.java:54]: moved to different test class and use JSON

File: utam-compiler/src/test/java/utam/compiler/helpers/ReturnTypeTests.java [hunk 2]
```diff
@@ -63,22 +58,4 @@ public void testReturnStringUnsupported() {
...
-  @Test
-  public void testAbstractReturnArrayEmpty() {
-    String expectedErr = getUnsupportedArrayTypeError("[ ]");
-    test("abstractReturnArrayEmpty", expectedErr);
-  }
-
-  @Test
-  public void testAbstractReturnArrayNonBasic() {
-    String expectedErr = getUnsupportedArrayTypeError("[ \"one\", \"two\" ]");
-    test("abstractReturnArrayNonBasic", expectedErr);
```
Comment (lizaiv77) [inline utam-compiler/src/test/java/utam/compiler/helpers/ReturnTypeTests.java:82]: no longer validated in this class

File: utam-compiler/src/test/java/utam/compiler/helpers/TypeUtilitiesTests.java [hunk 3]
```diff
@@ -232,25 +222,6 @@ public void testFromClassHashCode() {
...
-    assertThat(
-        Objects.requireNonNull(asBasicType("basicType", new String[] { "editable" }, false)).getSimpleName(),
-        is(equalTo("BasicTypeElement")));
-    assertThat(
-        Objects.requireNonNull(asBasicType("basicType", new String[] { "clickable" }, true)).getSimpleName(),
-        is(equalTo("GetBasicTypeElement")));
-    assertThat(asBasicType("basicType", new String[] { "unknown" }, false), is(nullValue()));
-    assertThat(asBasicType("basicType", null, false).getSimpleName(), is(equalTo("BasicTypeElement")));
-  }
-
```
Comment (lizaiv77) [inline utam-compiler/src/test/java/utam/compiler/helpers/TypeUtilitiesTests.java:253]: recreated as JSON tests

**Review by jimevanssfdc — APPROVED** [review]
LGTM

**Pull request merged**